### ISSUE TITLE
add verification step to check that large files have been successfully copied onto kube pods

### ISSUE THREAD

**Issue by jrhizor** [airbytehq/airbyte#8643]
We've seen some evidence that some large config files are partially copied or not copied at all when launching Kubernetes pods. 

see https://airbytehq.slack.com/archives/C02E3KDDRQE/p1638963725102900 for more info

We should separately confirm that files are completely copied before setting the success file.

**Comment by sherifnada** [comment]
@jrhizor could this be due to https://github.com/airbytehq/airbyte/issues/8625 ?

**Comment by jrhizor** [comment]
Yeah, maybe.

**Comment by harshithmullapudi** [comment]
Hey small update here, to test this out I checked the code https://github.com/kubernetes-client/java/blob/f68f17720e153bf5ddca6e63df70477c776e765f/util/src/main/java/io/kubernetes/client/Copy.java#L329

This looks to be like they do async copy which wouldn't be good in this case as FINISHED_UPLOADING is of 0 bytes and was causing lot of intermittent issues. But ideal solution is to use https://github.com/kubernetes-client/java/blob/90c4c1a012c66db963ebe8383d403a6c461ec284/util/src/main/java/io/kubernetes/client/Copy.java#L412 which means to update the package to 14.0.0 which I am sure how hard it is to. 

As I am not good with JAVA I went for something stupid 😢 🤣 and added this line

```
copy.copyFileToPod(namespace, podName, INIT_CONTAINER_NAME, contents, containerPath);
TimeUnit.SECONDS.sleep(2); //This line
```

`Before this line`: I have seen 32 syncs fail out of 100 for this reason
`After this line`: I have not seen this at all (ran about 500 syncs)

**Comment by jrhizor** [comment]
The synchronous call should work. It looks like the later one allows async copying and then just checking the future. I'm not sure why that would be more robust. We can try upgrading the client (should be easy).

@harshithmullapudi were you doing something to run a bunch of small syncs quickly? Is there a script or something I can just steal for testing this?

**Comment by cgardens** [comment]
@harshithmullapudi would you mind trying to update the client like @jrhizor mentioned and see if that fixes the problem?

**Comment by jrhizor** [comment]
@lmossman here's some additional context:
There's a problem where sometimes we see invalid/incomplete/(maybe missing, i'm not sure?) catalog files after the `KubePodProcess` initializes. 

When we initialize a "pod process", we run a sequence of `copyFileToPod` commands to copy each config file over. Then we upload a file called `FINISHED_UPLOADING`. We're copying this in the init container for the pod. The [entrypoint for the init container](https://github.com/airbytehq/airbyte/blob/master/airbyte-workers/src/main/java/io/airbyte/workers/process/KubePodProcess.java#L150) waits until the `FINISHED_UPLOADING` file is available. 

One hypothesis is that the `copyFileToPod` method isn't performing a synchronous copy like it should.

The first step though is to reliably reproduce the problem locally. Probably the easiest thing to try is to add something to `KubePodProcessIntegrationTest` that just stands up a new process that tries to transfer a large file file. Maybe doing this 10+ times in a loop will result in a high chance of failure. It's a possibility that this isn't actually the root cause of the problem, or that it's extremely difficult to reproduce. However it sounds like Harshith was able to reproduce this by running syncs. If a simplified version of the reproduction doesn't work, it's worth syncing with him to understand what kind of catalog he was syncing and how he was running syncs to reliably encounter the failure.

After that, the goal is of course to stabilize this. We may be using the Kube client api incorrectly, or there may be a way for us to check for transfer completeness before uploading the `FINISHED_UPLOADING` file and allowing the pod to run its main container.

**Comment by lmossman** [comment]
An update here: I am able to reliably reproduce this issue in a test I added to `KubePodProcessIntegrationTest` (see test [here](https://github.com/airbytehq/airbyte/pull/8880/files#diff-4776ad538cbe8b9807303d3f74b46b9ab07b6240cbd9053fb714cc95c45a23c1R215)). It looks like when the files being copied are large enough, some of the files are only partially copied.

For example, in a test run where I copied three ~1MB files, two of the files were fully copied but the third was only partially copied, despite the pod starting up and the `FINISHED_UPLOADING` file being copied in:
```
/config # ls -l
total 2508
-rw-r--r--    1 root     root             0 Dec 17 19:04 FINISHED_UPLOADING
-rw-r--r--    1 root     root       1010000 Dec 17 19:04 file0
-rw-r--r--    1 root     root       1010000 Dec 17 19:04 file1
-rw-r--r--    1 root     root        544384 Dec 17 19:04 file2
```

**Comment by lmossman** [comment]
Looking at the most recent version of the java kubernetes client, it looks like they added logic to create a future on the copy process and wait for it to complete before returning from `copyFileToPod`: https://github.com/kubernetes-client/java/blob/d4103b1bd5f1f6c82c65cd9f237307df44536a5e/util/src/main/java/io/kubernetes/client/Copy.java#L379-L410
as opposed to the version we are using, which does not create a future or block until it is complete: https://github.com/kubernetes-client/java/blob/f68f17720e153bf5ddca6e63df70477c776e765f/util/src/main/java/io/kubernetes/client/Copy.java#L352-L371

So the root cause is in fact that the version we are using is doing an async copy process which does not block until it completes. The solution here is to upgrade the version of the java kube client that we are using

**Comment by cgardens** [comment]
Any idea what the LOE of the upgrade is? If it is high, is there anything we can do to wait on the version we already have (obviously less ideal).

**Comment by lmossman** [comment]
Meant to post this earlier, but here is what I tried Friday evening:
1. Upgraded to the most recent version of the official java kube client. This did not work, as it would hang on `copyFileToPod`. This is an already-reported issue with the client: https://github.com/kubernetes-client/java/issues/1822 (no fix yet)
    - Based on [this comment](https://github.com/kubernetes-client/java/issues/1822#issuecomment-944509979) in the above issue, I tried using `copyFileToPodAsync()` with a while loop checking on `isDone()`, but this didnt work (files were still only partially copied)
    - I tried using the code in [this comment](https://github.com/kubernetes-client/java/issues/1822#issuecomment-900682240) that tries to brute force the bytes to be written out, but that also resulted in partially copied files.
2. I tried using `copyFileToPodAsync()` in the upgraded version, and then sleeping for 10 seconds before moving on to the next file. This actually resulted in the files being fully copied over. 
    - The issue here was that I would get the following error after the files were copied over, which has also been reported (and not fixed) in this issue: https://github.com/kubernetes-client/java/issues/1489
        ```
        2021-12-17 23:10:08 ERROR i.k.c.u.WebSocketStreamHandler(close):138 - {} - Error on flush
        java.io.IOException: Socket is closed!
	    at io.kubernetes.client.util.WebSocketStreamHandler$WebSocketOutputStream.flush(WebSocketStreamHandler.java:221) ~[client-java-14.0.0.jar:?]
	    at io.kubernetes.client.util.WebSocketStreamHandler.close(WebSocketStreamHandler.java:136) [client-java-14.0.0.jar:?]
	    at io.kubernetes.client.Exec$ExecProcess$1.close(Exec.java:542) [client-java-14.0.0.jar:?]
	    at io.kubernetes.client.Exec$ExecProcess$1.handleMessage(Exec.java:508) [client-java-14.0.0.jar:?]
        ```
    - Since the files were still being fully copied, I was thinking that I could get this method to work. However, when I tried this on GKE, I would get `WebSocket has closed` errors _during_ the copy, which would result in the files only being partial copied and the streams being closed.
3. I tried going back to using the `fabric8io` client for copying. This was successful on docker desktop. But when I tried on GKE, it successfully copied 90% of the 100MB files I used to test over, with only one file being partially copied.

Since the `fabric8` client was the closest one to successfully copying all of the large files over to GKE without any errors, I am hoping that I can somehow get it to wait and make sure it copies every byte. 

But I'm open to any other ideas or suggestions on this. Unfortunately it seems like both clients are broken in their own ways, so there is no ideal solution here.

**Comment by lmossman** [comment]
I'm trying out strategy #2 again from above, but using the existing client version that we are using today, and it appears that the `sleep` approach with this client version is actually working without issue, even on GKE. So if we can get this to just wait for all of the bytes to be copied over, I think this approach will work.

**Comment by lmossman** [comment]
I just tried the above approach for a 100MB file, and unfortunately it always only copies ~12MB of the file over, no matter how long I `sleep`

**Comment by lmossman** [comment]
One other data point here: using `kubectl cp` from the command line to copy a 100MB file to the GKE pod works flawlessly. So the above issues are only manifested in the java clients. This means that shelling out to do the copy may be an option here

**Comment by lmossman** [comment]
I tried out the `fabric8` client once more to see if I could add in some waiting or anything, and now it is throwing multiple errors when a large (i.e. >50MB) file gets uploaded. The file is fully uploaded, but even if I catch the errors and just log them, then it fails immediately on the next file that it tries to upload (even if I close and create a new client!). 

I just tried out executing a `kubectl cp` command in the java process to do the copy, and it worked flawlessly; files were fully copied over and it did not result in any errors. 

Even though it feels bad to execute a `kubectl` shell command from the java process to do this copy, given the state of these API's copy methods it seems like this might just be the most stable option.

**Comment by lmossman** [comment]
Last night, Charles and I were able to discover why only 12MB (or 16MB when we removed the `tar` step) was able to be copied using the official client. It is a limitation imposed by WebSocket, described here: https://github.com/square/okhttp/issues/3619

Now that we understand why the copying of files abruptly stops at this amount of data, we could just communicate that we have a limit of 10MB for any config files. 
However, there is still the issue of blocking - the current version of the official client that we use does not block on the copy to complete, but rather returns almost immediately. Adding a `sleep` here helps, but this is a brittle solution. A more robust solution would be to track how many bytes have been written into the container in real time, and only stop when all bytes have been written.

Charles and I decided that the above approach would be more effort to implement and would likely still be more brittle than just using `kubectl cp` directly, which flawlessly copies any file size and correctly blocks until every byte has been written into the container. Therefore, I am working on a PR to have this use `kubectl cp` to perform the copy.
